examples/riot/coap_channel/coap_channel.c

```diff
@@ -0,0 +1,284 @@
+  // the connection to us as established.
+
+  /* Client */
+  switch (self->state) {
+  case NETWORK_CHANNEL_STATE_CONNECTED:
+    return LF_OK;
+
+  case NETWORK_CHANNEL_STATE_OPEN:
+    if (!_send_coap_message(&self->remote, "/connect", _client_try_connect_callback)) {
+      LF_ERR(NET, "CoapChannel: try_connect: Failed to send CoAP message");
+      return LF_ERR;
+    }
+    return LF_IN_PROGRESS;
+
+  case NETWORK_CHANNEL_STATE_CONNECTION_IN_PROGRESS:
+    return LF_IN_PROGRESS;
+
+  case NETWORK_CHANNEL_STATE_DISCONNECTED:
+  case NETWORK_CHANNEL_STATE_LOST_CONNECTION:
+    self->state = NETWORK_CHANNEL_STATE_OPEN;
+    return LF_TRY_AGAIN;
+
+  case NETWORK_CHANNEL_STATE_UNINITIALIZED:
+    return LF_ERR;
+  }
+}
+
+static void CoapChannel_close_connection(NetworkChannel *untyped_self) {
+  LF_DEBUG(NET, "CoapChannel: Close connection");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  // TODO
+}
+
+static lf_ret_t CoapChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
+  LF_DEBUG(NET, "CoapChannel: Send blocking");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  // TODO
+}
+
+static void CoapChannel_register_receive_callback(NetworkChannel *untyped_self,
+                                                  void (*receive_callback)(FederatedConnectionBundle *conn,
+                                                                           const FederateMessage *msg),
+                                                  FederatedConnectionBundle *conn) {
+  LF_INFO(NET, "CoapChannel: Register receive callback");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+static void CoapChannel_free(NetworkChannel *untyped_self) {
+  LF_DEBUG(NET, "CoapChannel: Free");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  // TODO
+}
+
+void CoapChannel_ctor(CoapChannel *self, Environment *env, const char *remote_host) {
+  // Initialize global coap server it not already done
+  if (!_is_coap_initialized) {
+    _is_coap_initialized = true;
+
+    // Set environment
+    _env = env;
+
+    // Initialize coap server
+    gcoap_register_listener(&_listener);
+  }
+
+  // Super fields
+  self->super.open_connection = CoapChannel_open_connection;
+  self->super.try_connect = CoapChannel_try_connect;
+  self->super.close_connection = CoapChannel_close_connection;
+  self->super.send_blocking = CoapChannel_send_blocking;
+  self->super.register_receive_callback = CoapChannel_register_receive_callback;
+  self->super.free = CoapChannel_free;
+
+  // Concrete fields
+  self->receive_callback = NULL;
+  self->federated_connection = NULL;
+  self->state = NETWORK_CHANNEL_STATE_UNINITIALIZED;
+
+  // Convert host to udp socket
+  sock_udp_name2ep(&self->remote, remote_host);
+  if (self->remote.port == 0) {
+    self->remote.port = CONFIG_GCOAP_PORT;
+  }
```

Comment:
If user doesnt specify a port of the remote, then how can we address the particular CoapChannel at the remote (he might have several?)

Reply:
The CoAP channel uses the default CoAP port. This decision was made after it turned out to be very hard to have multiple CoAP servers on one device using different ports. The CoAP library in RIOT sets the server port based on that define :( 

But after I thought about this more deeply I think this is actually fine.
We have just one CoAP server yes, but we have multiple channels. When the server gets a request it passes it on to the correct channel based on the IP address of the other federate that the request originates from.